File: introHPC/09/thread_sync.cpp
#include <iostream>
#include <iomanip>
#include <thread>
#include <deque>
#include <vector>
#include <mutex>

using namespace std;

const long MAX = 10000;
const int nthreads = 2;

void worker(mutex &m, deque<long> &d, int threadNo, vector<long> &sumWorkers)
{
    while(1) {
	unique_lock<mutex> lck {m};
	if (d.empty()) break;
        long num = d.front();
        d.pop_front();
	lck.unlock();
        sumWorkers[threadNo] += num;
    }
}

auto main() -> int
{
    mutex m;
    deque<long> d(MAX);
    vector<long> sumWorkers(nthreads);
    thread threads[nthreads];
    for (long i=1; i<=MAX; i++) {
        d.push_back(i);
    }

    for (int i=0; i<nthreads; i++) {
	sumWorkers[i] = 0;
        threads[i] = thread(worker, ref(m), ref(d), i, ref(sumWorkers));
    }

    long sum = 0;
    for (int i=0; i<nthreads; i++) {
        threads[i].join();
        sum += sumWorkers[i];
    }
    cout << "Sum should be: " << setiosflags(ios::fixed) 
		<< setprecision(0) << MAX / 2.0 * (MAX +1) << endl;
    cout << "Sum calculate: " << sum << endl;

    return 0;
}

File: introHPC/09/thread_sync.py
import multiprocessing as mp
from functools import reduce

MAX = 10000
nthreads = 2

   
def worker(lock1, d, threadNo, sumWorkers):
    sum = 0
    while True:
        with lock1:
            if d.empty():
                break
            num = d.get(block=False)
        sum += num
    sumWorkers.put(sum)
    
if __name__ == '__main__':
   
    d = mp.Queue()
    sumWorkers = mp.Queue(nthreads)
    
    [d.put(i) for i in range(1, (MAX+1))]
    lock1 = mp.Lock();
    threads = [mp.Process(target=worker, args=(lock1, d, i, sumWorkers)) 
        for i in range(0, nthreads)]
    [t.start() for t in threads]
    [t.join() for t in threads]
    
    
    sum = 0
    print("Got = [", end='')
    while sumWorkers.empty() == False:
        num = sumWorkers.get()
        print("%d, " % num, end='')
        sum += num
    print("]")
    print("Sum should be: %d" % (MAX / 2.0 * (MAX+1)))
    print("Sum calculate: %d" % sum)